rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+
+For the below data, Encodes and Decodes are ops/sec so higher is better while size before and after compression lower is better. This list is far from exhaustive and specifcially lacks a serde_json comparison which is unfortunate but other benches out there do include them (though obviously a different benchmark makes cross comparison difficult). 
+
+| Library            | Encode | Decode All | Decode 1 | Update 1 | Size (bytes) | Size (Zlib) |
+|--------------------|--------|------------|----------|----------|--------------|-------------|
+| **Runtime Libs**   |        |            |          |          |              |             |
+| *NoProto*          |   1057 |       1437 |    47619 |    12195 |          208 |         166 |
+| Apache Avro        |    138 |         51 |       52 |       37 |          702 |         336 |
+| FlexBuffers        |    401 |        855 |    23256 |      264 |          490 |         309 |
+| JSON               |    550 |        438 |      544 |      396 |          439 |         184 |
+| BSON               |    115 |        103 |      109 |       80 |          414 |         216 |
+| MessagePack        |    135 |        222 |      237 |      119 |          296 |         187 |
+| **Compiled Libs**  |        |            |          |          |              |             |
+| Flatbuffers        |   1046 |      14706 |   250000 |     1065 |          264 |         181 |
+| Bincode            |   5882 |       8772 |     9524 |     4016 |          163 |         129 |
+| Protobuf           |    859 |       1140 |     1163 |      480 |          154 |         141 |
+| Prost              |   1225 |       1866 |     1984 |      962 |          154 |         142 |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks).  Some of these formats (JSON for example) still leaves us with a marginal concern around both dropped metadata as well as incidental exposure of a private data format to our users. The goal here is to hopefully solve for all of these things with whatever path is chosen.
+
+While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. As was [expressed by @lukesteensen](https://discord.com/channels/742820443487993987/746070604192415834/796425792875397120) in the #development channel in discord we have three specific concerns around the choice:
+
+-  Our internal data model itself (inclusive of our desire to not incidentally create _another_ observability data format)
```

Comment:
I agree that it'd be nice to leverage an existing format if it makes sense, but I don't think it is a hard requirement of internal vector <-> vector communication. For this, I think the priorities are:

* Performance
* Maintaining compatibility (e.g. if an old version of vector is sending to a newer one)
* Compatibility with a variety of transports (e.g. https://github.com/timberio/vector/issues/5124 mentions wanting to use HTTP)

Reply:
Not necessarily in that order. I think backwards compatibility is, or at least will become, the top priority, followed by performance and compatibility with different transports.